source/daplink/interface/target_flash.c

```diff
@@ -65,9 +67,40 @@ const flash_intf_t *const flash_intf_target = &flash_intf;
 
 static flash_func_t last_flash_func = FLASH_FUNC_NOP;
 
+//saved flash algo
+static program_target_t * current_flash_algo = NULL;
+
+//saved flash start from flash algo
+static uint32_t flash_start = 0;
+
+static program_target_t * get_flash_algo(uint32_t addr)
+{
+    region_info_t * flash_region = g_board_info.target_cfg->extra_flash;
+    
+    //check for the first flash area
+    if (addr >= g_board_info.target_cfg->flash_start && addr <= g_board_info.target_cfg->flash_end) {
+        flash_start = g_board_info.target_cfg->flash_start;
+        return g_board_info.target_cfg->flash_algo;
+    }
+    
+    for (; flash_region->start != 0 && flash_region->end != 0; ++flash_region) {
+        if (addr >= flash_region->start && addr <= flash_region->end) {
+            flash_start = flash_region->start; //save the flash start
+            if (flash_region->flash_algo) {
+                return flash_region->flash_algo;
+            }else{
+                break;
+            }
+        }
+    }
+    
+    //use the current flash algo by default
+    return current_flash_algo;
```

Comment:
Sorry if I wasn't clear… I meant that if the address doesn't match a flash algo, we should ultimately error out rather than trying to use an algo that may or may not work. So just return NULL from here and check it where called.